pallets/manta-sbt/src/lib.rs

```diff
@@ -0,0 +1,1029 @@
+            TransferPostError::<T>::Sender(err) => err.into(),
+            TransferPostError::<T>::Receiver(err) => err.into(),
+            TransferPostError::<T>::DuplicateMint => Self::DuplicateRegister,
+            TransferPostError::<T>::DuplicateSpend => Self::DuplicateSpend,
+            TransferPostError::<T>::InvalidProof => Self::InvalidProof,
+            TransferPostError::<T>::UnexpectedError(_) => Self::InternalLedgerError,
+        }
+    }
+}
+
+impl<T: Config> Pallet<T> {
+    /// Maximum Number of Updates per Shard (based on benchmark result)
+    const PULL_MAX_RECEIVER_UPDATE_SIZE: u64 = 32768;
+
+    /// Pulls receiver data from the ledger starting at the `receiver_indices`.
+    /// The pull algorithm is greedy. It tries to pull as many as possible from each shard
+    /// before moving to the next shard.
+    #[inline]
+    fn pull_receivers(
+        receiver_indices: [usize; MerkleTreeConfiguration::FOREST_WIDTH],
+        max_update_request: u64,
+    ) -> (bool, ReceiverChunk) {
+        let mut more_receivers = false;
+        let mut receivers = Vec::new();
+        let mut receivers_pulled: u64 = 0;
+        let max_update = if max_update_request > Self::PULL_MAX_RECEIVER_UPDATE_SIZE {
+            Self::PULL_MAX_RECEIVER_UPDATE_SIZE
+        } else {
+            max_update_request
+        };
+
+        for (shard_index, utxo_index) in receiver_indices.into_iter().enumerate() {
+            more_receivers |= Self::pull_receivers_for_shard(
+                shard_index as u8,
+                utxo_index,
+                max_update,
+                &mut receivers,
+                &mut receivers_pulled,
+            );
+            // NOTE: If max capacity is reached and there is more to pull, then we return.
+            if receivers_pulled == max_update && more_receivers {
+                break;
+            }
+        }
+        (more_receivers, receivers)
+    }
+
+    /// Pulls receiver data from the shard at `shard_index` starting at the `receiver_index`,
+    /// pushing the results back to `receivers`.
+    #[inline]
+    fn pull_receivers_for_shard(
+        shard_index: u8,
+        receiver_index: usize,
+        max_update: u64,
+        receivers: &mut ReceiverChunk,
+        receivers_pulled: &mut u64,
+    ) -> bool {
+        let max_receiver_index = (receiver_index as u64) + max_update;
+        for idx in (receiver_index as u64)..max_receiver_index {
+            if *receivers_pulled == max_update {
+                return Shards::<T>::contains_key(shard_index, idx);
+            }
+            match Shards::<T>::try_get(shard_index, idx) {
+                Ok(next) => {
+                    *receivers_pulled += 1;
+                    receivers.push(next);
+                }
+                _ => return false,
+            }
+        }
+        Shards::<T>::contains_key(shard_index, max_receiver_index)
+    }
+
+    /// Pulls sender data from the ledger starting at the `sender_index`.
```

Comment:
This comment is wrong now. Maybe worth scanning this page again .

Reply:
Underlying Function is unneeded too